modules/renter.go

```diff
@@ -50,12 +54,96 @@ type Allowance struct {
 // DownloadInfo provides information about a file that has been requested for
 // download.
 type DownloadInfo struct {
-	SiaPath     string    `json:"siapath"`
-	Destination string    `json:"destination"`
-	Filesize    uint64    `json:"filesize"`
-	Received    uint64    `json:"received"`
-	StartTime   time.Time `json:"starttime"`
-	Error       string    `json:"error"`
+	SiaPath     string         `json:"siapath"`
+	Destination DownloadWriter `json:"destination"`
+	Filesize    uint64         `json:"filesize"`
+	Received    uint64         `json:"received"`
+	StartTime   time.Time      `json:"starttime"`
+	Error       string         `json:"error"`
+}
+
+// DownloadWriter provides an interface which all output writers have to implement.
+type DownloadWriter interface {
+	WriteAt(b []byte, off int64) (int, error)
```

Comment:
John and I talked a bit and we feel that `DownloadWriter` is the wrong abstraction for what we want to achieve. The limitations of `WriteAt` don't fit the constraints that we have to work with when doing massively parallel downloads under memory constrained situations, especially if we are trying to eek out every ounce of performance and cost-efficiency.

What I think instead should be done is have the downloader only operate one chunk at a time, and then when the downloader is finished fetching that chunk, it'll return it.

`DownloadChunk(index, offset uint64) ([]byte, error)`

Then the caller is responsible for managing what happens with the data afterwords. And at some point I can see us extending the DownloadChunk function to have priority as well or something (streaming would want a more consistent latency at the very least, and maybe even to get download-boosted on every chunk vs. just the final ones).

I think the `WriteAt` function is just going to cause lots of headaches as we try to optimize around different situations. Seems much more flexible to grab one chunk at a time and then let the caller handle optimizing, ordering, memory constraints, etc.